File: benchmarks/complex/hashmap.cpp
#include <cstdio>
#include <vector>
using std::vector;
static long lookup(const vector<long>& keys, const vector<long>& vals, long cap, long key) {
    long idx = (key * 2654435761L) % 2147483648L % cap, steps = 0;
    while (steps < cap) { long k = keys[idx]; if (k == 0) return 0; if (k == key) return vals[idx]; idx++; if (idx >= cap) idx = 0; steps++; }
    return 0;
}
static void insert(vector<long>& keys, vector<long>& vals, long cap, long key, long val) {
    long idx = (key * 2654435761L) % 2147483648L % cap;
    while (keys[idx] > 0) { idx++; if (idx >= cap) idx = 0; }
    keys[idx] = key; vals[idx] = val;
}
static void remove_(vector<long>& keys, long cap, long key) {
    long idx = (key * 2654435761L) % 2147483648L % cap, steps = 0;
    while (steps < cap) { long k = keys[idx]; if (k == 0) return; if (k == key) { keys[idx] = -1; return; } idx++; if (idx >= cap) idx = 0; steps++; }
}
int main() {
    long cap = 1048576; vector<long> keys(cap, 0), vals(cap, 0); long n = 400000;
    for (long i = 1; i <= n; i++) insert(keys, vals, cap, i, i * 3);
    long checksum = 0;
    for (long q = 1; q <= 800000; q++) { long key = q * 7919 % 800000 + 1; checksum = (checksum + lookup(keys, vals, cap, key)) % 1000000007L; }
    for (long i = 2; i <= n; i += 2) remove_(keys, cap, i);
    for (long q = 1; q <= 800000; q++) { long key = q * 7919 % 800000 + 1; checksum = (checksum + lookup(keys, vals, cap, key)) % 1000000007L; }
    printf("%ld\n", checksum);
}

File: benchmarks/complex/hashmap.rs
fn lookup(keys: &[i64], vals: &[i64], cap: i64, key: i64) -> i64 {
    let mut idx = (key * 2654435761) % 2147483648 % cap;
    let mut steps = 0;
    while steps < cap {
        let k = keys[idx as usize];
        if k == 0 { return 0; }
        if k == key { return vals[idx as usize]; }
        idx += 1; if idx >= cap { idx = 0; }
        steps += 1;
    }
    0
}
fn insert(keys: &mut [i64], vals: &mut [i64], cap: i64, key: i64, val: i64) {
    let mut idx = (key * 2654435761) % 2147483648 % cap;
    while keys[idx as usize] > 0 { idx += 1; if idx >= cap { idx = 0; } }
    keys[idx as usize] = key; vals[idx as usize] = val;
}
fn remove(keys: &mut [i64], cap: i64, key: i64) {
    let mut idx = (key * 2654435761) % 2147483648 % cap;
    let mut steps = 0;
    while steps < cap {
        let k = keys[idx as usize];
        if k == 0 { return; }
        if k == key { keys[idx as usize] = -1; return; }
        idx += 1; if idx >= cap { idx = 0; }
        steps += 1;
    }
}
fn main() {
    let cap = 1048576i64;
    let mut keys = vec![0i64; cap as usize];
    let mut vals = vec![0i64; cap as usize];
    let n = 400000i64;
    let mut i = 1; while i <= n { insert(&mut keys, &mut vals, cap, i, i * 3); i += 1; }
    let mut checksum = 0i64;
    let mut q = 1; while q <= 800000 { let key = q * 7919 % 800000 + 1; checksum = (checksum + lookup(&keys, &vals, cap, key)) % 1000000007; q += 1; }
    i = 2; while i <= n { remove(&mut keys, cap, i); i += 2; }
    q = 1; while q <= 800000 { let key = q * 7919 % 800000 + 1; checksum = (checksum + lookup(&keys, &vals, cap, key)) % 1000000007; q += 1; }
    println!("{}", checksum);
}
